Firestore/Source/API/FSTUserDataReader.mm

```diff
@@ -491,26 +557,100 @@ - (void)parseSentinelFieldValue:(FIRFieldValue *)fieldValue context:(ParseContex
           other.project_id(), other.database_id(), _databaseID.project_id(),
           _databaseID.database_id(), context.FieldDescription());
     }
-    return FieldValue::FromReference(_databaseID, reference.key);
+    return [self encodeReference:_databaseID key:reference.key];
 
   } else {
     ThrowInvalidArgument("Unsupported type: %s%s", NSStringFromClass([input class]),
                          context.FieldDescription());
   }
 }
 
-- (std::vector<FieldValue>)parseArrayTransformElements:(NSArray<id> *)elements {
+- (google_firestore_v1_Value)encodeNullValue {
```

Comment:
Note: in C++, I'd suggest making these free functions (they don't need to access anything in the class). I'm not sure what is considered best practices in Objective-C in this regard, but it seems like similar reasoning should apply. The advantage of free functions is that they cannot access private data -- this makes them a little easier to reason about and prevents them from coming to rely on the internal state of the class by accident.

Optionally, you could move these functions into a pure C++ file.

Reply:
`parseReference` accesses `self.databaseID`, which means `parseReference`, `parseArray` and `parseObject` cannot be free functions. For consistency, I chose to keep the other functions as is.

Reply:
You could pass it as a parameter, but it's probably not worth the trouble.